Add data container and pm.set_data

jmloyola opened the pull request:
This work is related to the issue (#2056 ). Below is a summary of the changes:
- Create data container class (`pm.Data`) that wraps the `theano.compile.SharedVariable` class and let the model be aware of its inputs and outputs.
- Add data variables from model to graphviz digraph as observed variables.
- Add new API `model.predict()` that allows you to predict output data given a new set of input data.
- Add new API `model.refit()` that allows you to refits the model using new data.
- Add tests for new the data container class and the new methods.
- Update `bayesian_neural_network_advi.ipynb` notebook with the new API.

I want to thank @aloctavodia for the help given by answering questions about pymc3 codebase. :smile: 

Any comments or suggestions are more than welcome.

jmloyola committed:
Add data container, model.predict and model.refit

ColCarroll commented:
Took a quick look, and this looks very, very interesting and tricky! Will make more substantive technical comments, but thank you for working through this.

ColCarroll commented on pymc3/data.py:
```diff
@@ -385,3 +386,61 @@ def align_minibatches(batches=None):
+
+        # To draw the node for this variable in the graphviz Digraph we need
+        # its shape.
+        shared_object.dshape = tuple(shared_object.shape.eval())
```
we should add a new shape for data sources.

junpenglao replied:
Also, be careful of the potential shape change for theano.shared variables. You should use `draw_values` that contains better check for shape: https://github.com/pymc-devs/pymc3/blob/master/pymc3/distributions/distribution.py#L502-L517

ColCarroll commented on pymc3/model.py:
```diff
@@ -1054,6 +1054,113 @@ def _repr_latex_(self, name=None, dist=None):
+
+        ppc = pm.sample_posterior_predictive(trace=trace, model=self, **kwargs)
+
+        for variable_name, original_value in dict_original_values.items():
```
cool -- so calling this method does not change the state of the `Data` or `model`. 

I think this should be wrapped in a `try/finally`, since we tend to try to handle a `StopExecution` during sampling?

```python
try:
    ppc = pm.sample_posterior_predictive(...)
finally:
    for variable_name, original_value in dict_original_values.items():
        ...
```

ColCarroll commented on pymc3/model.py:
```diff
@@ -1054,6 +1054,113 @@ def _repr_latex_(self, name=None, dist=None):
 
     __latex__ = _repr_latex_
 
+    def predict(self, trace, dict_new_values, **kwargs):
```
it sort of feels like this should _be_ `sample_posterior_predictive`, but I am ok merging like this first, and thinking about how to do that.

ColCarroll reviewed (commented):
Left a few thoughts - interested in other opinions since this is a big change.

twiecki commented on pymc3/data.py:
```diff
@@ -385,3 +386,61 @@ def align_minibatches(batches=None):
                 rng.seed()
+
+
+class Data():
```
don't need parantheses.

twiecki commented:
This is really cool, thanks!

I wonder about the API (which I'm aware I originally proposed). We have a pretty consistent functional API rather than a method-based one which this API is in conflict with. It would be more in line to instead do:
```python
with model:
    pm.predict(trace, new_data)
```

From there, it's pretty close to
```python
with model:
    pm.sample_posterior_predictive(trace, update=new_data)
```
like @ColCarroll suggested. 

Or maybe this instead:
```python
with model:
    pm.update_data(new_data)
    pm.sample_posterior_predictive(trace)
```
which would be the most general and perhaps least surprising.

jmloyola commented:
Thank you all for your comments!

Do you think the function `pm.predict` / `pm.sample_posterior_predictive` should retain the original data after execution or not?

If it should, I think it is straightforward to implement it using this API:
```python
with model:
    pm.sample_posterior_predictive(trace, update=new_data)
```

But I don't know how we will do it with this one:
```python
with model:
    pm.update_data(new_data)
    pm.sample_posterior_predictive(trace)
```
Should we go back to the original value after the call to the function `pm.sample_posterior_predictive(trace)`? What if someone whants to `pm.sample_posterior_predictive(trace)` many times with the same `new_data`?

twiecki commented:
@jmloyola The idea is that `pm.update_data()` would in-place change the data in the shared vars. It'd be the users job to keep track of that.

Curious what others think in terms of API @junpenglao @ColCarroll @fonnesbeck

aloctavodia commented:
Updating the data using
`pm.sample_posterior_predictive(trace, update=new_data)` is simpler than having `pm.update_data()`. Also no need to retain old data, you can always pass data, old or new, to the `update` argument.

twiecki commented:
I'm fine with that. @jmloyola does that make sense?

jmloyola commented:
Yes, I will work on this and all the suggestions previously mentioned.

jmloyola committed:
Use functional API instead of one based in methods

aloctavodia commented:
@twiecki do you think this is ready to merge?

twiecki commented:
What's the purpose of `refit`? I don't think that one is super useful asit'd be just one extra line to do `date.update`.

twiecki commented:
I would also add a NB dedicated to this new functionality and add a line in the release-notes.

twiecki commented:
An option I would like better is to add the same `update` kwarg to `sample()` and `fit()` but not introduce a new `refit` function.

But I still don't like adding these kwargs to all functions where might want to allow a user to change data. I still think a `pm.update_data()` that replaces the data of a model is the cleanest way to introduce this functionality, as it'd be the only thing we need to add to the API.

ferrine commented:
`update` would be a misleading keyword argument. What about more explicit ones like `data` or `set_data`?

twiecki commented:
@ferrine What do you think of the consideration of adding a `set_data` kwarg to  `sample`, `fit` and `sample_posterior_perdictive`, vs a single new function `pm.set_data`?

ferrine commented:
I would prefer `model.set_data` with a possibility to use it as a context manage that returns the state back ofc.

Context manager is totally optional. I can imagine memory leak when used wrong.

ferrine commented:
New arguments will call the new function, more ways to do the same thing. But it is easier to use. `pm.set_data` would be more consistent with the rest api and probably the way to go

twiecki commented:
I'm quite convinced `pm.set_data()` is the way to go here.

jmloyola commented:
Thanks you all for your comments! :smile: 

If I understood correctly, I should:
- [x] Remove the method `pm.refit()`.
- [x] Remove the parameter _update_ from the function `sample_posterior_predictive()`.
- [x] Add the function `pm.set_data()` that should receive a dictionary with variables' name to update and the values to use. This function should only `set_value()` for each variable to update, checking for type errors obviously.
- [x] Add a jupyter notebook showing the use of the data container and the new `pm.set_data()`.
- [x] Add a line in the release-notes mentioning the new function.

Do you agree? @twiecki @ferrine @fonnesbeck @junpenglao @ColCarroll @aloctavodia

twiecki commented:
@jmloyola Exactly!

jmloyola committed:
remove refit function, remove update parameter from sample_posterior_predictive and add set_data function

jmloyola committed:
Update RELEASE-NOTES.md

ferrine commented on pymc3/model.py:
```diff
@@ -1055,6 +1100,43 @@ def _repr_latex_(self, name=None, dist=None):
+
+    .. code:: ipython
+        >>> pm.set_data(new_data={'x': [5,6,9]}, model=model)
+        >>> # Another way to do the same thing using the model context is:
```
I would say "using model itself"

ericmjl replied:
Minor english correction, @ferrine:

`using the model itself`

twiecki commented on pymc3/data.py:
```diff
@@ -12,7 +12,8 @@
     'Minibatch',
-    'align_minibatches'
+    'align_minibatches',
+    'Data'
```
```suggestion
    'Data',
```

twiecki commented on pymc3/data.py:
```diff
@@ -385,3 +386,61 @@ def align_minibatches(batches=None):
+        ...     with model:
+        ...         traces.append(pm.sample())
+    """
+    def __new__(self, name, value):
```
Why can't this happen in `__init__`?

twiecki committed:
add trailing comma in pymc3/data.py

Co-Authored-By: jmloyola <jmloyola@outlook.com>

jmloyola committed:
Minor english correction

twiecki commented on pymc3/model.py:
```diff
@@ -920,6 +921,50 @@ def makefn(self, outs, mode=None, *args, **kwargs):
+        Set the value of `x` to predict on new data.
+
+        .. code:: ipython
+            >>> pm.set_data(new_data={'x': [5,6,9]}, model=model)
```
I would change this to also use the `with model` syntax which we're  using almost everywhere else (and remove the note about calling `model.set_data()`).

twiecki commented on pymc3/model.py:
```diff
@@ -920,6 +921,50 @@ def makefn(self, outs, mode=None, *args, **kwargs):
                                    accept_inplace=True,
                                    mode=mode, *args, **kwargs)
 
+    def set_data(self, new_data):
```
Does this need to be a method? Can't it just be a function and operate on the passed in `model`? I really want to try to keep the API consistent here, and there is no `model.sample()` or other user-facing methods.

jmloyola committed:
Remove model.set_data() and update pm.set_data's docstring

twiecki commented on pymc3/model.py:
```diff
@@ -1055,6 +1056,53 @@ def _repr_latex_(self, name=None, dist=None):
+
+    .. code:: ipython
+        >>> with model:
+        ...     pm.set_data(new_data={'x': [5,6,9]})
```
```suggestion
        ...     pm.set_data({'x': [5,6,9]})
```

twiecki commented on pymc3/data.py:
```diff
@@ -385,3 +386,61 @@ def align_minibatches(batches=None):
+        The name for this variable
+    value
+        A value to associate with this variable
+
```
I would also add a note to check out `pm.set_data()` so that users know the intended usage.

jmloyola committed:
fix failing test

jmloyola committed:
add note to check set_data function inside the Data class docstring and clean examples

jmloyola committed:
Add a jupyter notebook showing the use of the data container and the new pm.set_data()

jmloyola committed:
Fix typo in notebook

jmloyola committed:
Add tests to increase test coverage

twiecki commented:
NB looks great!

* "a observed" -> "an observed"
* "Note that we used a theano SharedVariable method to get the value of the variable. This is because our variable is actually a SharedVariable." You could add a `type(data)` call to show that.
* print(pm.set_data.__doc__) -> help(pm.set_data)
* "Lenghts (heights) of newborn / toddlers girls as a function of the age (in months)" -> "Applied example: height of toddlers as a function of age"
* "This charts" -> "These charts"
* "health and related policies" -> "health policies"
* "newborn / toddlers" -> "newborn / toddler"
* "of the age" -> "of age"
* replace all `sd` -> `sigma`, e.g. `pm.Normal('α', sd=10)`
* "daughter has two weeks" -> "daughter is two weeks (=0.5 months) old"
* `sample_ppc` -> `sample_posterior_predictive` (also in text)
* "as this will return samples of conditioned" -> "as this will return samples conditioned"
* "predictions for for"
* "values of , and $0.5$ months (the value he cares)" - value he cares *about*  and missing value
* 
![image](https://user-images.githubusercontent.com/674200/54824744-190ed880-4cac-11e9-80f2-bf71ad904078.png) add more descriptive x and y labels.

Also make sure to include the NB in the docs and then link to them from the help of pm.Data.

jmloyola commented:
I added some tests to increase the test coverage and added the notebook showing the use of the data container and the new `pm.set_data()`.

The notebook can be seen [here](https://github.com/jmloyola/pymc3/blob/add_data_container/docs/source/notebooks/data_container.ipynb). If you find any mistake or typo or want me to add something to the notebook, just tell me and I will try to fix it.

jmloyola committed:
Fix typos in notebook

jmloyola committed:
Add link to notebook in docstring

twiecki commented:
Well, this is a high quality PR -- great work!

I assume you ran the tests locally and they pass?

jmloyola commented:
I'm running the test suite now. It might take some time in my machine :grin:

jmloyola commented:
I think the test suite is way too big for my computer. It took almost four hours to run all and several of the tests failed with the error `OSError: [Errno 12] Cannot allocate memory`.

The good news is that TravisCI says that everything it's OK. Is that good enough? Does TravisCI run all the test suite?

ericmjl commented:
@jmloyola yes, it's good enough that Travis tests pass! I'll leave it to someone else with more experience to hit that big green button, if there's no other opinions.

twiecki commented:
Thanks @jmloyola! this is great stuff.
